### PR TITLE
Proxy actions PR updates

### PR DESCRIPTION
**Pull request by brianmcmichael**
## Updates

* some nitpicky things
* add underscores to internal functions for consistency
* adds a test of an implementation replacement
* optimization in proxy actions
* modify some lines to get under 120 chars (did not go with 80 char limit @godsflaw may have something to say about that)

## Open questions, potential issues:

* Does `exit()` need a call to `validate`? It seems like we might be able to draw dai against collateral down to the charter-enforced upper bound, and then `exit` the remaining collateral to end up down at the system-collateralization ratio.

* We may want to add a single function to set up a new user into the whitelist. Otherwise we're looking at 4 separate `file` calls to add a new whitelisted user. Seems like we should provide a single function for this.

* Potential issue in `CharterManager.sol`:

```
    if (dart > 0 && _gate == 1) {
        require(tab <= uline[ilk][u], "CharterManager/user-line-exceeded");
    }
```
This seems to prevent partial payback by a user if their debt ceiling has been reduced to `0`.

* The MIP implies that there would be minimum qualifying sizes for vaults. This implementation seems to only enforce a minimum vault size by way of system `dust` but does not enforce a minimum deposit otherwise.

### CODE REVIEW

**Comment by talbaneth** [comment]
@brianmcmichael see some responses to the `Open questions, potential issues` part:

* Does exit() need a call to validate? It seems like we might be able to draw dai against collateral down to the charter-enforced upper bound, and then exit the remaining collateral to end up down at the system-collateralization ratio.

_I don't think there is an issue. `exit()` does not take out collateral from the vault, just gems. The peace validation is for the vault values `ink` and `art`, which can not be affected by `exit()`_

* We may want to add a single function to set up a new user into the whitelist. Otherwise we're looking at 4 separate file calls to add a new whitelisted user. Seems like we should provide a single function for this.

_Yes. Since whitelisting is authed I think the correct place would be DssExecLib (same as for new collateral in the vat or clip)._

* Potential issue in CharterManager.sol:

    if (dart > 0 && _gate == 1) {
        require(tab <= uline[ilk][u], "CharterManager/user-line-exceeded");
    }
This seems to prevent partial payback by a user if their debt ceiling has been reduced to 0.

_In dai paybeck dart is negative, so there shouldn't be an issue._

* The MIP implies that there would be minimum qualifying sizes for vaults. This implementation seems to only enforce a minimum vault size by way of system dust but does not enforce a minimum deposit otherwise.

_Yes. The minimum qualifying size is an off-chain agreement, enforced by governance._

**Review by talbaneth — APPROVED** [review]
LGTM

**Comment by brianmcmichael** [comment]
> I don't think there is an issue. exit() does not take out collateral from the vault, just gems. The peace validation is for the vault values ink and art, which can not be affected by exit()

Makes sense I think we assumed this. Maybe we should add a test to ensure this case.

> Yes. Since whitelisting is authed I think the correct place would be DssExecLib (same as for new collateral in the vat or clip).

I don't think I agree here. This is going to be a common operation on this contract. A single auth'ed function to set up a new user eliminates three extra external calls and the overhead around checking the `auth` and `what` status. These are tightly coupled values that `must` be in place for a user to be properly configured, so a user initialization function here ensures everything is handled in one place.

> In dai paybeck dart is negative, so there shouldn't be an issue.

Good point. :+1: We might want to add a test to demonstrate this.

> Yes. The minimum qualifying size is an off-chain agreement, enforced by governance.

How does this work in a permissionless way?

**Comment by talbaneth** [comment]
> Makes sense I think we assumed this. Maybe we should add a test to ensure this case.

Done in the original PR (https://github.com/makerdao/dss-charter/pull/6)

> I don't think I agree here. This is going to be a common operation on this contract. A single auth'ed function to set up a new 
user eliminates three extra external calls and the overhead around checking the `auth` and `what` status. These are tightly 
coupled values that _must_ be in place for a user to be properly configured, so a user initialization function here ensures everything is handled in one place.

We should probably have a policy for this. Let's discuss offline.
 
> Good point. +1 We might want to add a test to demonstrate this.

Done in the original PR (https://github.com/makerdao/dss-charter/pull/6)

> How does this work in a permissionless way?

It's not :).
For institutional vaults (gate=1) it is enforced by goverment through the migrating spell and through possible sanctions like removing DC or raising SFs/OFs.
For permissionless vaults (gate=0) we assume there is no initial mandatory size.
OFC everything can be upgraded.

File: src/CharterManager.sol [hunk 1]
```diff
@@ -117,7 +117,7 @@ contract CharterManagerImp {
 
     // --- Implementation Storage ---
     mapping (address => address) public proxy; // UrnProxy per user
-    mapping (address => mapping (address => uint256)) public can;
+    mapping (address => mapping(address => uint256))  public can;
```
Comment (gbalabasquer) [inline src/CharterManager.sol:120]: I think actually this line was fine as it was before.

Comment (brianmcmichael) [reply src/CharterManager.sol:120]: see line 123

Comment (brianmcmichael) [reply src/CharterManager.sol:120]: actually, this aligns with the others, but yeah, doesn't match the first mapping format. I'll fix the other ones.

Comment (brianmcmichael) [reply src/CharterManager.sol]: 47c904bf30b5620d497d44cee000cd0eff97d4f2

Comment (brianmcmichael) [reply src/CharterManager.sol]: solidity docs are pretty loose on whether it's

`mapping(k => v)`

or 

`mapping (k => v)`

So I guess we just need to pick one and be consistent.

**Review by talbaneth — APPROVED** [review]

**Review by gbalabasquer — APPROVED** [review]
LGTM

**Pull request merged**